service/matching/matcher_test.go

```diff
@@ -328,47 +329,97 @@ func (t *MatcherTestSuite) TestForwardingWhenBacklogIsEmpty() {
 }
 
 func (t *MatcherTestSuite) TestAvoidForwardingWhenBacklogIsOld() {
-	intruptC := make(chan struct{})
+	t.childConfig.MaxWaitForPollerBeforeFwd = dynamicconfig.GetDurationPropertyFn(20 * time.Millisecond)
+
+	interruptC := make(chan struct{})
 	// forwarding will be triggered after t.cfg.MaxWaitForPollerBeforeFwd() so steps in this test should finish sooner.
-	maxWait := t.childConfig.MaxWaitForPollerBeforeFwd() / 2
+	maxWait := t.childConfig.MaxWaitForPollerBeforeFwd() * 2 / 3
 
 	// poll forwarding attempt happens when there is no backlog.
 	// important to make this empty poll to set the last poll timestamp to a recent time
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
+	forwardPoll := make(chan struct{})
+	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+		func(ctx context.Context, in *matchingservice.PollWorkflowTaskQueueRequest, opts ...grpc.CallOption) (*matchingservice.PollWorkflowTaskQueueResponse, error) {
+			forwardPoll <- struct{}{}
+			return &matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest
+		})
 	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	t.Eventually(
-		func() bool {
-			return t.childMatcher.timeSinceLastPoll() < time.Second
-		}, maxWait, time.Millisecond)
+	select {
```

Comment:
I recently pushed in a fix for a flaky which was caused since it was not waiting for the poll to be forwarded. I had used a waitGroup as a sync construct. Curious to hear why you chose going ahead with a channel over a waitGroup? Moreover, should we be keeping a standard for this unit test file (i.e - waitGroup/Channels)?

Reply:
I used a channel here so I could select on it with the context timeout and not wait for the overall test timeout. I think we have a "wait for waitgroup with timeout" in a library somewhere, so that could also work. I don't know if there's a strong reason to prefer one over the other, though I would tend to use a waitgroup for more than one thing and a channel for one.